[peripheral] change addon interface to new C++ style

AlwinEsch opened the pull request:
<!--- Provide a general summary of your change in the Title above -->

## Description
This change also the peripheral addon system to use C++ class as
interface to Kodi.

There is only the interface itself changed, the functions style is still
the same as before with the use of PeripheralUtils.h (kodi_peripheral_utils.hpp).
My experience for the system is not good enough to change also the list functions
to use direct a vector on addon (not sure how I can test every function).

Maybe can be this improved from @garbear who have more experience.

The test of them here was good and has played some games :-) .

<!--- Describe your change in detail -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

notspiff commented on xbmc/peripherals/bus/virtual/PeripheralBusAddon.cpp:
```diff
@@ -457,23 +402,26 @@ void CPeripheralBusAddon::UpdateAddons(void)
+          CSingleExit exit(m_critSection);
+          bCreated = newAddon->CreateAddon();
+        }
+fprintf(stderr, "-----3- %s - %i\n", __PRETTY_FUNCTION__, bCreated);
```
Nah

AlwinEsch replied:
Uppsss, readed several times the commit and stil this test print inside, SORRY

notspiff replied:
No worries. last week i sent our customers code which spewed 'hore' five times on every run of the simulator :) (add a w in front to translate the norwegian..) proud moment ;)

garbear commented on xbmc/peripherals/bus/virtual/PeripheralBusAddon.cpp:
```diff
@@ -514,9 +462,9 @@ void CPeripheralBusAddon::PromptEnableAddons(const ADDON::VECADDONS& disabledAdd
+    [](const BinaryAddonBasePtr& addonInfo)
   {
-    return std::static_pointer_cast<CPeripheralAddon>(addon)->HasButtonMaps();
+    return addonInfo->Type(ADDON::ADDON_PERIPHERALDLL)->GetValue("@provides_buttonmaps").asBoolean();
```
Any way to not scatter XML tags in this function?

AlwinEsch replied:
Not direct, it can be maybe a static function added on ./xbmc/peripherals/addons/PeripheralAddon.h who do this with pass of addonInfo as value.

This way was thought (on agile) to replace cpluff and allow a read of addon.xml value without creation of related addon class, in this case CPeripheralAddon. Also used here to allow read of values of several different instance types without creation of them.

garbear replied:
Makes sense, maybe a static function HasButtonMaps() that takes the xml parameter?

AlwinEsch replied:
I add them

garbear committed:
[peripheral] Cosmetic: Fix doxy indentation

garbear committed:
[inputstream] Cosmetic: Fix doxy indentation

garbear commented:
@AlwinEsch https://github.com/AlwinEsch/kodi/pull/30

ksooo reviewed (commented):
Looking good. Nothing to moan this time.

AlwinEsch commented:
Is updated and use now `static inline ...` inside CPeripheralAddon to get extension values.

tamland commented on xbmc/addons/binary-addons/BinaryAddonManager.h:
```diff
@@ -63,6 +63,24 @@ namespace ADDON
+     * @return true if enabled
+     */
+    bool IsAddonEnabled(const std::string& addonId, const TYPE &type = ADDON_UNKNOWN);
+
```
This is not something that should be added as public methods on BinaryAddonManager (same with GetDisabledAddonInfos). AddonMgr still have the authority over which addons are installed an enabled so all of these might give wrong/outdated answers.

tamland commented on xbmc/peripherals/bus/virtual/PeripheralBusAddon.cpp:
```diff
@@ -439,8 +383,9 @@ void CPeripheralBusAddon::UpdateAddons(void)
-  PeripheralAddonVector newAddons = GetEnabledAddons();
-  std::transform(newAddons.begin(), newAddons.end(), std::inserter(newIds, newIds.end()), GetPeripheralAddonID);
+  BinaryAddonBaseList newAddons;
+  CServiceBroker::GetBinaryAddonManager().GetAddonInfos(newAddons, true, ADDON_PERIPHERALDLL);
```
Like here. This doesn't work. CPeripheralBusAddon listens to events from addon manager, so there is no guarantee BinaryAddonManager will have the correct list at this point. And then we are out of sync.

AlwinEsch replied:
This and the other is thought as replacement for the CBinaryAddonCache and to remove them in future, both does mostly the same.

AlwinEsch replied:
With use of GetAddon direct is everything lost why them both are present.

tamland replied:
> This and the other is thought as replacement for the CBinaryAddonCache and to remove them in future, both does mostly the same.

Ok. No need to fix right now then. I see that the set_difference etc. below isn't correct either (doesn't not account for updates) so this whole method will need to be changed at some point.

AlwinEsch commented:
@garbear OK from your side to bring in?

jenkins build this please

AlwinEsch commented:
jenkins build this please

garbear commented:
OK from my side. Concerning `PeripheralUtils.h`, it would be good to convert this to the class system that you've implemented for the rest of the add-ons. This header was just my filler until we had a C++ interface. But this can happen anytime... I'll be in the mountains for the next week or two so I can look into this when I return.

AlwinEsch committed:
[peripheral] change addon interface to new C++ style

This change also the peripheral addon system to use C++ class as
interface to Kodi.

There is only the interface itself changed, the functions style is still
the same as before with the use of PeripheralUtils.h (kodi_peripheral_utils.hpp).
My experience for the system is not good enough to change also the list functions
to use direct a vector on addon (not sure how I can test every function).

The test of them here was good and has played some games :-) .

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
jenkins build this please
